oioioi/problems/tests/test_tag_proposals.py

```diff
@@ -0,0 +1,337 @@
+        response = self.client.get(self.get_query_url({'query': 'dynam'}))
+        self.assertEqual(response.status_code, 200)
+        self.assertContains(response, 'Programowanie dynamiczne')
+        self.assertNotContains(response, 'dp')
+        self.assertNotContains(response, 'Dynamic programming')
+
+        response = self.client.get(self.get_query_url({'query': 'greedy'}))
+        self.assertEqual(response.status_code, 200)
+        self.assertContains(response, b'Zach\u0142anny')
+        self.assertNotContains(response, 'Greedy')
+
+        # Use a byte string in the query to ensure a proper url encoding.
+        response = self.client.get(self.get_query_url({'query': 'ZAchłan'}))
+        self.assertEqual(response.status_code, 200)
+        self.assertContains(response, b'Zach\u0142anny')
+        self.assertNotContains(response, 'Greedy')
+
+        response = self.client.get(self.get_query_url({'query': 'longest'}))
+        self.assertEqual(response.status_code, 200)
+        self.assertContains(
+            response, b'Najd\u0142u\u017cszy wsp\u00f3lny podci\u0105g rosn\u0105cy'
+        )
+        self.assertNotContains(response, 'Longest common increasing subsequence')
+        self.assertNotContains(response, 'lcis')
+
+        response = self.client.get(self.get_query_url({'query': 'lcis'}))
+        self.assertEqual(response.status_code, 200)
+        self.assertContains(
+            response, b'Najd\u0142u\u017cszy wsp\u00f3lny podci\u0105g rosn\u0105cy'
+        )
+        self.assertNotContains(response, 'Longest common increasing subsequence')
+        self.assertNotContains(response, 'lcis')
+
+
+class TestSaveProposals(TestCase):
+    fixtures = [
+        'test_users',
+        'test_problem_search',
+        'test_algorithm_tags',
+        'test_difficulty_tags',
+    ]
+    url = reverse('save_proposals')
+
+    def test_save_proposals_view(self):
+        problem = Problem.objects.get(pk=0)
+        user = User.objects.get(username='test_admin')
+
+        self.assertEqual(AlgorithmTagProposal.objects.count(), 0)
+        self.assertEqual(AggregatedAlgorithmTagProposal.objects.count(), 0)
+        self.assertEqual(DifficultyTagProposal.objects.count(), 0)
+        self.assertEqual(AggregatedDifficultyTagProposal.objects.count(), 0)
+
+        response = self.client.post(
+            self.url,
+            {
+                'tags[]': ["Dynamic programming", "Knapsack problem"],
+                'difficulty': '  \r    \t\n        Easy   \n     \t  ',
+                'user': 'test_admin',
+                'problem': '0',
+            },
+        )
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(AlgorithmTagProposal.objects.count(), 2)
+        self.assertEqual(AggregatedAlgorithmTagProposal.objects.count(), 2)
+        self.assertEqual(DifficultyTagProposal.objects.count(), 1)
+        self.assertEqual(AggregatedDifficultyTagProposal.objects.count(), 1)
+        self.assertTrue(
```

Comment:
looks very complicated, maybe add some utility function that reads tag proposals and returns a dictionary with {name: amount}